### DISCUSSION TITLE
Error: Write failed. Insufficient number of disks online (*errors.errorString)

### DISCUSSION THREAD

**Discussion by sathishkumar-p** [minio/minio#17619] [Q&A] [answered]
Hello, 
I am running minio two-node cluster with 750GB, I am frequently getting the following error.

Error: Write failed. Insufficient number of disks online (*errors.errorString)

API: PutObject(bucket=-test, object=inputs/Trade.xlsx)
Time: 07:11:26 UTC 07/10/2023
DeploymentID: 8a6ea546-1d6d-481f-9a4a-2fe0fe98b32b
RequestID: 17707012D51D6FEE
RemoteHost: 192.168.139.209
Host: minio-0.minio-headless.storage.svc.cluster.local:9000
UserAgent: MinIO (Linux; amd64) minio-java/8.4.6
Error: Write failed. Insufficient number of disks online (*errors.errorString)
       7: cmd/erasure-encode.go:104:cmd.(*Erasure).Encode()
       6: cmd/erasure-object.go:902:cmd.erasureObjects.putObject()
       5: cmd/erasure-object.go:710:cmd.erasureObjects.PutObject()
       4: cmd/erasure-sets.go:929:cmd.(*erasureSets).PutObject()
       3: cmd/erasure-server-pool.go:868:cmd.(*erasureServerPools).PutObject()
       2: cmd/object-handlers.go:1798:cmd.objectAPIHandlers.PutObjectHandler()
       1: net/http/server.go:2047:http.HandlerFunc.ServeHTTP()
Client http://minio-1.minio-headless.storage.svc.cluster.local:9000/minio/storage/data-1/v43 online
Client http://minio-1.minio-headless.storage.svc.cluster.local:9000/minio/storage/data-0/v43 online
Client http://minio-1.minio-headless.storage.svc.cluster.local:9000/minio/lock/v7 online

## Your Environment
* Version used (`minio --version`): RELEASE.2023-05-18T00-05-36Z
* Server setup and configuration: 2 node cluster
* Operating System and version (`uname -a`): kubernetes

**Comment by klauspost** [comment]
Something is flaky in your setup. With only 2 nodes, you leave yourself vulnerable to this.

**Reply by sathishkumar-p** [reply]
Hi @klauspost ,
The reason why keep these two nodes is because of less usage. what's your recommendation?

**Reply by sathishkumar-p** [reply]
Error: Write failed. Insufficient number of drives online (*errors.errorString)

we have 4 drivers and 2 server

**Comment by vyom-soft** [comment]
Hello, I am also seeing similar error:
https://minio.slack.com/archives/C3NDUB8UA/p1688992061601199
 
 We are on multi node cluster (RedHat). With rook-ceph as dynamic storage.
  Deployed Minio via Helm Chart, is in distributed mode and set the storageClass: blockstorage
  logs shows following error:
  Error: Write failed. Insufficient number of drives online (*errors.errorString)        Same error is show when the storageClass is set to file   
  system. 

Any advice please.

Thank you.

```
		API: SYSTEM()
	Time: 12:21:16 UTC 07/10/2023
	DeploymentID: e0986c7c-b40f-4f0f-ac6a-57034c20d078
	Error: Write failed. Insufficient number of drives online (*errors.errorString)
	       9: internal/logger/logger.go:258:logger.LogIf()
	       8: cmd/erasure-encode.go:112:cmd.(*Erasure).Encode()
	       7: cmd/erasure-object.go:1190:cmd.erasureObjects.putObject()
	       6: cmd/erasure-object.go:954:cmd.erasureObjects.PutObject()
	       5: cmd/erasure-sets.go:767:cmd.(*erasureSets).PutObject()
	       4: cmd/erasure-server-pool.go:953:cmd.(*erasureServerPools).PutObject()
	       3: cmd/config-common.go:83:cmd.saveConfigWithOpts()
	       2: cmd/config-common.go:88:cmd.saveConfig()
	       1: cmd/data-usage.go:52:cmd.storeDataUsageInBackend()

	API: SYSTEM()
	Time: 12:21:16 UTC 07/10/2023
	DeploymentID: e0986c7c-b40f-4f0f-ac6a-57034c20d078
	Error: Storage resources are insufficient for the write operation .minio.sys/tmp/		
	92592b11-476c-4d1f-9cbc-085ab4f863a7/568b1b2b-6cda-472a-98fc-db25d24d07f5/part.1 (cmd.InsufficientWriteQuorum)
	       3: internal/logger/logger.go:258:logger.LogIf()
	       2: cmd/data-scanner.go:228:cmd.runDataScanner()
	       1: cmd/data-scanner.go:80:cmd.initDataScanner.func1()
```

**Reply by ravindk89** [reply]
See the answer above - in this case you are giving MinIO drives attached over the network, and those devices are themselves Ceph presented through an additional layer of Rook.

MinIO recommends local storage. If you do use NAS/DAS or similar networked storage, you have to ensure the downstream storage service has really tight network and service availability. If any of those things are flaky, or (if I had to guess) the interface medium doesn't really give the same behavior as a local drive on MinIO's IO pattern, you're going to have trouble.

Your second error indicates some other sort of storage problem. If these rook-ceph volumes aren't formatted as XFS you should at least try to fix that.

**Reply by vyom-soft** [reply]
@ravindk89 Thank you for the support 
I did the following changes in the config.
```
  replicaCount: 2
  drivesPerNode: 2 
```
Therefore, the total drives, The total number of drives should be greater than 4 to guarantee erasure coding!(replica X drivesPerNode = 4)
With these changes, the above err disappeared and I see new err.
```
```
<pre><font color="#26A269"><b></b></font>:<font color="#12488B"><b>~</b></font>$ kubectk get sts smoketest-minio -n smoketest
NAME                   READY   AGE
smoketest-minio   1/2     17m
</pre>

```
```
<pre>$ kubectl get pvc -n smoketest
NAME                                                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
data-0-smoketest-minio-0                        Bound    pvc-c8036616-4c01-448c-aee6-6eb6bad7822d   2Gi        RWO            ceph-filesystem   17m
data-0-smoketest-minio-1                        Bound    pvc-f7245f98-0be6-49d7-bd2e-a53791a2ab69   2Gi        RWO            ceph-filesystem   17m
data-1-smoketest-minio-0                        Bound    pvc-e3317d35-02b9-4ffd-8d66-877851f2726d   2Gi        RWO            ceph-filesystem   17m
data-1-smoketest-minio-1                        Bound    pvc-36b12c7c-eeb8-4948-8347-f3121f7a3f88   2Gi        RWO            ceph-filesystem   17m
</pre>

```


API: SYSTEM()
Time: 17:01:06 UTC 07/10/2023
Error: Marking smoketest-minio-1.smoketest-minio-headless.smoketest.svc.cluster.local:9000 offline temporarily; caused by Post "http://smoketest-minio-1.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/storage/data-0/v49/readall?disk-id=&file-path=format.json&volume=.minio.sys": dial tcp 10.233.75.117:9000: connect: connection refused (*fmt.wrapError)
       8: internal/logger/logonce.go:118:logger.(*logOnceType).logOnceIf()
       7: internal/logger/logonce.go:149:logger.LogOnceIf()
       6: internal/rest/client.go:259:rest.(*Client).Call()
       5: cmd/storage-rest-client.go:164:cmd.(*storageRESTClient).call()
       4: cmd/storage-rest-client.go:551:cmd.(*storageRESTClient).ReadAll()
       3: cmd/format-erasure.go:391:cmd.loadFormatErasure()
       2: cmd/format-erasure.go:327:cmd.loadFormatErasureAll.func1()
       1: internal/sync/errgroup/errgroup.go:123:errgroup.(*Group).Go.func1()
```
```
Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Client 'http://smoketest-minio-1.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/lock/v7' re-connected in 2m47.310873746s
Client 'http://smoketest-minio-1.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/storage/data-0/v49' re-connected in 2m47.432479099s
Client 'http://smoketest-minio-1.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/storage/data-1/v49' re-connected in 2m47.468517917s
Waiting for all MinIO sub-systems to be initialized.. lock acquired
Automatically configured API requests per node based on available memory on the system: 870
All MinIO sub-systems initialized successfully in 2m47.520452932s
MinIO Object Storage Server
Copyright: 2015-2023 MinIO, Inc.
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Version: DEVELOPMENT.2023-03-24T21-41-23Z (go1.19.7 linux/amd64)

Status:         4 Online, 0 Offline. 
API: http://localhost:9000 
Console: http://10.233.90.62:9001 http://127.0.0.1:9001   

Documentation: https://min.io/docs/minio/linux/index.html
Stopping on service signal
```

Can you please direct where the configuration is getting wrong ?

Volumes are properly XFS formatted, otherwise rook-ceph will not be healthy at all. And other components which uses rook-ceph, (redhat default to XFS) and other STS apps are in stable state.

Thank you.

**Reply by ravindk89** [reply]
I don't know anything about rook/ceph, so I cannot really troubleshoot issues happening using their volumes under the hood. 

> dial tcp 10.233.75.117:9000: connect: connection refused 

This looks like a networking issue of some sort.

**Reply by harshavardhana** [reply]
2 nodes is unsupported for production deployments @vyom-soft if you use it - you are generally on your own.

**Reply by vyom-soft** [reply]
Thank you for the inputs.

I am on a cluster with 6 worker nodes & 2 Master nodes. So the min requirement of having 4 nodes is fulfilled.
I now have
  replicaCount: 2
  drivesPerNode: 4
 ∴ total drives= 8

Why is this now  -> Err : /data-0 returned drive not found 

```
API: SYSTEM()
Time: 07:52:09 UTC 07/11/2023
DeploymentID: 338686b9-69ba-4327-abdb-4c8bfcdd2ed4
Error: Drive: http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/data-0 returned drive not found (*fmt.wrapError)
       endpoint="http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/data-0"
       3: internal/logger/logger.go:248:logger.LogAlwaysIf()
       2: cmd/prepare-storage.go:50:cmd.glob..func7.1()
       1: cmd/erasure-sets.go:220:cmd.(*erasureSets).connectDisks.func2()
Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
...
Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock
Client 'http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/storage/data-0/v49' re-connected in 5m4.240220199s
Client 'http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/lock/v7' re-connected in 5m4.187093712s
Client 'http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/storage/data-1/v49' re-connected in 5m4.362716745s
Waiting for all MinIO sub-systems to be initialized.. lock acquired
Automatically configured API requests per node based on available memory on the system: 697
All MinIO sub-systems initialized successfully in 5m4.311634515s
MinIO Object Storage Server
Copyright: 2015-2023 MinIO, Inc.
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Version: DEVELOPMENT.2023-03-24T21-41-23Z (go1.19.7 linux/amd64)

Status:         4 Online, 0 Offline. 
API: http://localhost:9000 
Console: http://10.xxxx.xx.118:9001 http://127.0.0.1:9001   

Documentation: https://min.io/docs/minio/linux/index.html
Stopping on service signal
```

**Reply by klauspost** [reply]
`Storage resources are insufficient for the write operation .minio.sys/tmp/` is an error returned from the filesystem, so there isn't much we can do here.

I suggest you look into why your filesystem is returning errors.

**Reply by vyom-soft** [reply]
@klauspost Thank you for the reply.  

I don't think this is the case of insufficient for the write operation, File system is okay. 

There is new error shown, pls. see  my new reply above.

```
I am trying to understand/find why mino is able to  write & read inf fron data-1 but why not from data-0

Error: Drive: http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/data-0 returned drive not found (*fmt.wrapError)
       endpoint="http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/data-0"
       3: internal/logger/logger.go:248:logger.LogAlwaysIf()
       2: cmd/prepare-storage.go:50:cmd.glob..func7.1()
       1: cmd/erasure-sets.go:220:cmd.(*erasureSets).connectDisks.func2()
```

**Reply by klauspost** [reply]
I can't dig deeper into your problem. We do not support ceph setups.

**Reply by vyom-soft** [reply]
Fresh err

API: SYSTEM()
Time: 11:27:47 UTC 07/11/2023
Error: Marking smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000 offline temporarily; caused by Post "http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/minio/storage/data-0/v49/readall?disk-id=&file-path=format.json&volume=.minio.sys": dial tcp 10.233.90.54:9000: connect: connection refused (*fmt.wrapError)
       8: internal/logger/logonce.go:118:logger.(*logOnceType).logOnceIf()
       7: internal/logger/logonce.go:149:logger.LogOnceIf()
       6: internal/rest/client.go:259:rest.(*Client).Call()
       5: cmd/storage-rest-client.go:164:cmd.(*storageRESTClient).call()
       4: cmd/storage-rest-client.go:551:cmd.(*storageRESTClient).ReadAll()
       3: cmd/format-erasure.go:391:cmd.loadFormatErasure()
       2: cmd/format-erasure.go:327:cmd.loadFormatErasureAll.func1()
       1: internal/sync/errgroup/errgroup.go:123:errgroup.(*Group).Go.func1()

API: SYSTEM()
Time: 11:27:47 UTC 07/11/2023
DeploymentID: 338686b9-69ba-4327-abdb-4c8bfcdd2ed4
Error: Drive: http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/data-1 returned drive not found (*fmt.wrapError)
       endpoint="http://smoketest-minio-0.smoketest-minio-headless.smoketest.svc.cluster.local:9000/data-1"
       3: internal/logger/logger.go:248:logger.LogAlwaysIf()
       2: cmd/prepare-storage.go:50:cmd.glob..func7.1()
       1: cmd/erasure-sets.go:220:cmd.(*erasureSets).connectDisks.func2()

**Comment by ravindk89** [comment, marked as answer]
2 x 4 = 8 drives. With default EC parity you have quorum of 4 drives, or one server. Write Quorum when parity is 1/2 drives in the set is M+1 or 5 drives. So if one node goes down, you cannot write.

Any transient network issues are going to cause errors like this. If a node is flaky, if it restarts, etc. basically guarantees you lose write quorum.

This is why our documentation recommends at least 4 nodes as a minimum production setup - you at least protect yourself from a single flaky node or bad network link from being a show stopper.

**Reply by ravindk89** [reply]
If you want to dig further you can look at the pod and host logs and see if there are issues around DNS resolution, networking, etc. That would be the best place to start.

**Reply by sathishkumar-p** [reply]
Hello @ravindk89 ,
'the documentation recommends at least 4 nodes as a minimum production setup' . What is the recommendation for a number of drivers?
Also please share the link for minio recommendation step for disturbed

**Reply by vyom-soft** [reply]
In the _helpers.tpl  of helm chart, it is mentioned as
{{- $totalDrives := mul $replicaCount $drivesPerNode }}
minio: total drives
    The total number of drives should be greater than 4 to guarantee erasure coding!
    Please set a combination of nodes, and drives per node that match this condition.

**Reply by ravindk89** [reply]
@sathishkumar-p we have guidance [here](https://min.io/docs/minio/linux/operations/checklists/hardware.html#minio-hardware-checklist), but in general a 4 Servers x 4 Drives / Server is the baseline for production. That gets you an absolute minimum level of availability, resiliency, and distributed performance.

- [Deploy MinIO: Multi-Node Multi-Drive (Distributed)](https://min.io/docs/minio/linux/operations/install-deploy-manage/deploy-minio-multi-node-multi-drive.html)
- [Erasure Coding Core Concepts](https://min.io/docs/minio/linux/operations/concepts/erasure-coding.html)

**Reply by ravindk89** [reply]
For anything more specific regarding production setups, I have to point you at [SUBNET](https://min.io/pricing). Architecture and deployment review/validation falls under our commercial offering, as it is fact-specific and requires dedicated engineering resources to ensure the best possible outcome.
